### PR TITLE
Change API response with `Expect`

### PR DESCRIPTION
**Pull request by javiersanchz**
|Related issue|
|---|
| #21935  |

## Description

The API responses were modified when they contain the `Expect` header.

The corresponding test was added for the new function

Without Expect header:
```console
root@wazuh-master:/# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer $TOKEN"   
{
   "title": "Unauthorized",
   "detail": "No authorization token provided"
}
```
With Expect 100-continue:
```console
root@wazuh-master:/# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer $TOKEN" -H "Expect: 100-continue"
"100-continue"
```

With Different Expect:
```console
root@wazuh-master:/# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer test_token" -H "Expect: Test_header"
{
   "title": "error : 417",
   "detail": "Expectation Failed."
}
```

## Tests

```console
(unittest-env) wazuh@javier:~/Git/wazuh$ PYTHONPATH=/home/wazuh/Git/wazuh/api:/home/wazuh/Git/wazuh/framework python3 -m pytest api/api/test/test_error_handler.py
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.12, pytest-7.3.1, pluggy-1.4.0
rootdir: /home/wazuh/Git/wazuh/api/api
configfile: pytest.ini
plugins: anyio-4.3.0, aiohttp-1.0.4, trio-0.8.0, html-2.1.1, metadata-3.1.0, asyncio-0.18.1, tavern-1.23.5
asyncio: mode=auto
collected 31 items                                                                                                                                                                                                

api/api/test/test_error_handler.py ...............................                                                                                                                                          [100%]

========================================================================================= 31 passed, 9 warnings in 0.63s ==========================================================================================
```

### CODE REVIEW

File: api/api/error_handler.py [hunk 1]
```diff
@@ -55,6 +55,42 @@ def _cleanup_detail_field(detail: str) -> str:
     return ' '.join(str(detail).replace("\n\n", ". ").replace("\n", "").split())
 
 
+async def handle_expect_header(request: ConnexionRequest, exc: ContentSizeExceeded = None) -> ConnexionResponse:
+    """Handler for 'Expect' header.
```
Comment (GGP1) [inline api/api/error_handler.py:59]:
```suggestion
    """Handler for the 'Expect' HTTP header.
```

File: api/api/error_handler.py [hunk 2]
```diff
@@ -55,6 +55,42 @@ def _cleanup_detail_field(detail: str) -> str:
...
+
+    Returns
+    -------
+    Response
+        HTTP Response returned to the client.
+    """
+    problem = {
+       "title": "error : 417",
+       "detail": "Expectation Failed."
+    }
```
Comment (GGP1) [inline api/api/error_handler.py:76]: I think this should be

```suggestion
    problem = {
       "title": "Expectation failed",
       "detail": "Unknown Expect",
       "error": 417
    }
```

but we should use `WazuhError` like in the rest of the code instead of hardcoding it. I'm not sure if raising an error here will be catched though.

File: api/api/error_handler.py [hunk 3]
```diff
@@ -55,6 +55,42 @@ def _cleanup_detail_field(detail: str) -> str:
...
+
+    if "Expect" in request.headers:
```
Comment (GGP1) [inline api/api/error_handler.py:78]: Use single quotes, same for the rest of the file

```suggestion
    if 'Expect' in request.headers:
```

File: api/api/error_handler.py [hunk 4]
```diff
@@ -55,6 +55,42 @@ def _cleanup_detail_field(detail: str) -> str:
...
+            return json_response(data=problem, pretty=request.query_params.get('pretty', 'false') == 'true',
+                                 status_code=417, content_type=ERROR_CONTENT_TYPE)
+        
+        elif exc and expect_value == "100-continue":
+            return json_response(data=problem, pretty=request.query_params.get('pretty', 'false') == 'true',
+                                 status_code=417, content_type=ERROR_CONTENT_TYPE)
+            
+        else:
+            return json_response(data="100-continue", pretty=request.query_params.get('pretty', 'false') == 'true',
+                                 status_code=200)
```
Comment (GGP1) [inline api/api/error_handler.py:91]:
```suggestion
        continue_msg = '100-continue'
        data = continue_msg
        status_code = 200
        content_type = None

        if expect_value != continue_msg or exc and expect_value == continue_msg:
            data = problem
            status_code = 417
            content_type = ERROR_CONTENT_TYPE
            return json_response(data=problem, pretty=request.query_params.get('pretty', 'false') == 'true',
                                 status_code=417, content_type=ERROR_CONTENT_TYPE)
                                 
        return json_response(data=data, pretty=request.query_params.get('pretty', 'false') == 'true',
                                 status_code=status_code, content_type=content_type)
```

**Review by GGP1 — CHANGES_REQUESTED** [review]

File: api/api/error_handler.py [hunk 5]
```diff
@@ -55,6 +55,46 @@ def _cleanup_detail_field(detail: str) -> str:
...
+    Returns
+    -------
+    Response
+        HTTP Response returned to the client.
+    """
+    problem = {
+       "title": "Expectation failed",
+       "detail": "Unknown Expect",
+       "error": 417
+    }
```
Comment (GGP1) [inline api/api/error_handler.py:77]: This could be defined inside the `if expect_value != continue_msg or (exc and expect_value == continue_msg):` condition because it's not used elsewhere.

Also, can we use `WazuhError` like the rest of the code instead of harcoding the values?

File: api/api/error_handler.py [hunk 6]
```diff
@@ -55,6 +55,46 @@ def _cleanup_detail_field(detail: str) -> str:
...
+
+        continue_msg = '100-continue'
+        data = continue_msg
+        status_code = 200
+        content_type = None
+
+        if expect_value != continue_msg or (exc and expect_value == continue_msg):
+            data = problem
+            status_code = 417
+            content_type = ERROR_CONTENT_TYPE
```
Comment (GGP1) [inline api/api/error_handler.py:90]: These variables are not being used

```suggestion
```

File: api/api/error_handler.py [hunk 7]
```diff
@@ -55,6 +55,46 @@ def _cleanup_detail_field(detail: str) -> str:
...
+       "error": 417
+    }
+
+    if 'Expect' in request.headers:
+        expect_value = request.headers["Expect"].lower()
+
+        continue_msg = '100-continue'
+        data = continue_msg
+        status_code = 200
+        content_type = None
```
Comment (GGP1) [inline api/api/error_handler.py:85]: Since we are just returning the values if there is an error, we could omit declaring the variables and use them directly when it suceeds (line 94)

```py
return json_response(data='100-continue', pretty=request.query_params.get('pretty', 'false') == 'true',
                            status_code=200)
```

**Review by GGP1 — CHANGES_REQUESTED** [review]

File: api/api/error_handler.py (hunk 5 above)
Comment (javiersanchz) [reply api/api/error_handler.py]: https://github.com/wazuh/wazuh/issues/21935#issuecomment-1991758569

**Review by GGP1 — APPROVED** [review]
LGTM

File: api/api/error_handler.py [hunk 8]
```diff
@@ -55,6 +55,43 @@ def _cleanup_detail_field(detail: str) -> str:
     return ' '.join(str(detail).replace("\n\n", ". ").replace("\n", "").split())
 
 
+async def handle_expect_header(request: ConnexionRequest, exc: ContentSizeExceeded = None) -> ConnexionResponse:
```
Comment (Selutario) [inline api/api/error_handler.py:58]: For consistency reasons, since all other handler's function name end in `_handler`:
```suggestion
async def expect_header_error_handler(request: ConnexionRequest, exc: ContentSizeExceeded = None) -> ConnexionResponse:
```

Probably, this won't receive a `ContentSizeExceeded` either but a `WazuhError`.

File: api/api/error_handler.py [hunk 9]
```diff
@@ -55,6 +55,43 @@ def _cleanup_detail_field(detail: str) -> str:
...
+
+    if 'Expect' in request.headers:
+        expect_value = request.headers["Expect"].lower()
+
+        continue_msg = '100-continue'
+        data = continue_msg
+        status_code = 200
+        content_type = None
+
+        if expect_value != continue_msg or (exc and expect_value == continue_msg):
```
Comment (Selutario) [inline api/api/error_handler.py:82]: This error handler cannot be the one checking whether the `Expect` header is correct. Notice that this function will only be called when a 417 error is raised in another part of the framework. Therefore, if the header value is being checked here, it means that it is not being checked elsewhere and as a consequence, 417 will not be raised and this function will not be called. 

In fact, this is not working right now:
```
# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer $TOKEN" -H "Expect: Test_header"
{
   "data": {
      "title": "Wazuh API REST",
      "api_version": "4.9.0",
      "revision": 40900,
      "license_name": "GPL 2.0",
      "license_url": "https://github.com/wazuh/wazuh/blob/v4.9.0/LICENSE",
      "hostname": "wazuh-master",
      "timestamp": "2024-03-21T12:10:37Z"
   },
   "error": 0
}
```

We should create a new middleware which checks if the `Expect` header is included, if `100-continue` is contained, and probably, in case a `Content-length` header was also included, whether its value is bigger than the one defined in `api_conf['max_upload_size']`. This way, clients will ask/expect the API whether a payload is too big. If the API responds 417, it means that said content-length should not be sent because it would be refused.

In summary, the new middleware should raise a 417 error if any of the above mentioned conditions are unmet. Then, the API will catch the 417 error (`app.add_error_handler(417, error_handler.expect_header_error_handler)` in `wazuh-apid.py`) and will call this handler, which should format the response as it is being done below.

**Review by Selutario — CHANGES_REQUESTED** [review]
Although this looks good overall, some changes are required yet:

File: api/api/error_handler.py [hunk 10]
```diff
@@ -55,6 +55,28 @@ def _cleanup_detail_field(detail: str) -> str:
...
+    Returns
+    -------
+    Response
+        HTTP Response returned to the client.
+    """
+    problem = {
+        "title": "Expectation failed",
+        "detail": "Unknown Expect",
+        "error": 417
+    }
```
Comment (Selutario) [inline api/api/error_handler.py:75]: The detail field could be different depending on whether the user included a unknown expect or the Content-Length was exceeded:

```suggestion
    problem = {
        "title": "Expectation failed"
    }

    if exc.detail:
        problem['detail'] = exc.detail

File: api/api/error_handler.py [hunk 11]
```diff
@@ -55,6 +55,28 @@ def _cleanup_detail_field(detail: str) -> str:
...
+    return json_response(data=problem, pretty=request.query_params.get('pretty', 'false') == 'true',
+                        status_code=417, content_type=ERROR_CONTENT_TYPE)
```
Comment (Selutario) [inline api/api/error_handler.py:77]:
```suggestion
    return json_response(data=problem, pretty=request.query_params.get('pretty', 'false') == 'true',
                         status_code=exc.status, content_type=ERROR_CONTENT_TYPE)
```

File: api/api/middlewares.py [hunk 12]
```diff
@@ -257,3 +258,39 @@ async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -
...
+            expect_value = request.headers["Expect"].lower()
+            
+            if expect_value != '100-continue':
+                raise ExpectFailedException(status=417, title="Expectation failed", detail="Unknown Expect")
+            
+            if 'Content-Length' in request.headers:
+                content_length = int(request.headers["Content-Length"])
+                max_upload_size = default_api_configuration["max_upload_size"]
+                if content_length > max_upload_size:
+                    raise ExpectFailedException(status=417, title="Expectation failed", detail="Unknown Expect")
```
Comment (Selutario) [inline api/api/middlewares.py:293]: The detail field can be improved:
```suggestion
                    raise ExpectFailedException(status=417, title="Expectation failed",
                                                detail=f"Maximum content size limit ({max_upload_size}) exceeded "
                                                       f"({content_length} bytes read)")
```

File: api/api/middlewares.py [hunk 13]
```diff
@@ -257,3 +258,39 @@ async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -
...
+                
+        response = await call_next(request)
+        return response
```
Comment (Selutario) [inline api/api/middlewares.py:296]: Add empty line at the end:

```suggestion
        return response

```

File: api/api/test/test_middlewares.py [hunk 14]
```diff
@@ -305,3 +308,24 @@ async def test_check_block_ip_middleware(endpoint, method, call_check, mock_req)
...
+    call_next_mock = AsyncMock(return_value=response)
+
+    if expect_value != '100-continue':
+        with pytest.raises(ExpectFailedException):
+            await middleware.dispatch(mock_request, call_next_mock)
+        call_next_mock.assert_not_called()
+    else:
+        returned_response = await middleware.dispatch(mock_request, call_next_mock)
+        call_next_mock.assert_called_once_with(mock_request)
+        assert returned_response == response
```
Comment (Selutario) [inline api/api/test/test_middlewares.py:331]:
```suggestion
        assert returned_response == response

```

**Review by Selutario — CHANGES_REQUESTED** [review]
Just some minor changes:

**Review by Selutario — APPROVED** [review]
LGTM. The failing checks are related to missing deps.

**Comment by Selutario** [comment]
Unknown expect:
```
# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer $TOKEN" -H "Expect: Test_header" 
{
   "title": "Expectation failed",
   "detail": "Unknown Expect"
}
```

Bigger Content-Length than allowed:
```
# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer $TOKEN" -H "Expect: 100-continue" -H "Content-Length: 99999999"
{
   "title": "Expectation failed",
   "detail": "Maximum content size limit (10485760) exceeded (99999999 bytes read)"
}
```

Meet expectations:
```
# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Authorization: Bearer $TOKEN" -H "Expect: 100-continue"
{
   "data": {
      "title": "Wazuh API REST",
      "api_version": "4.9.0",
      "revision": 40900,
      "license_name": "GPL 2.0",
      "license_url": "https://github.com/wazuh/wazuh/blob/v4.9.0/LICENSE",
      "hostname": "wazuh-master",
      "timestamp": "2024-04-04T10:39:35Z"
   },
   "error": 0
}
```

No bearer token included:
```
# curl -k -X GET "https://localhost:55000/?pretty=true" -H "Expect: Test_header" 
{
   "title": "Unauthorized",
   "detail": "No authorization token provided"
}
```
